extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergKeyValuePartitionedLayout.java

```diff
@@ -3,168 +3,69 @@
         partitionColumnTypes = Arrays.stream(partitionColumns)
                 .map(colName -> TypeUtils.getBoxedType(tableDef.getColumn(colName).getDataType()))
                 .toArray(Class<?>[]::new);
-
-        this.cache = new HashMap<>();
     }
 
+    @Override
     public String toString() {
-        return IcebergFlatLayout.class.getSimpleName() + '[' + table.name() + ']';
+        return IcebergKeyValuePartitionedLayout.class.getSimpleName() + '[' + table.name() + ']';
     }
 
     @Override
-    public synchronized void findKeys(@NotNull final Consumer<IcebergTableLocationKey> locationKeyObserver) {
+    IcebergTableLocationKey keyFromDataFile(DataFile df, URI fileUri) {
         final Map<String, Comparable<?>> partitions = new LinkedHashMap<>();
-        try {
-            // Retrieve the manifest files from the snapshot
-            final List<ManifestFile> manifestFiles = currentSnapshot.allManifests(fileIO);
-            for (final ManifestFile manifestFile : manifestFiles) {
-                // Currently only can process manifest files with DATA content type.
-                Assert.eq(manifestFile.content(), "manifestFile.content()",
-                        ManifestContent.DATA, "ManifestContent.DATA");
-                final ManifestReader<DataFile> reader = ManifestFiles.read(manifestFile, fileIO);
-                for (DataFile df : reader) {
-                    final URI fileUri = FileUtils.convertToURI(df.path().toString(), false);
-                    final IcebergTableLocationKey locationKey = cache.computeIfAbsent(fileUri, uri -> {
-                        final PartitionData partitionData = (PartitionData) df.partition();
-                        for (int ii = 0; ii < partitionColumns.length; ++ii) {
-                            final Object value = partitionData.get(ii);
-                            if (value != null && !value.getClass().isAssignableFrom(partitionColumnTypes[ii])) {
-                                throw new TableDataException("Partitioning column " + partitionColumns[ii]
-                                        + " has type " + value.getClass().getName()
-                                        + " but expected " + partitionColumnTypes[ii].getName());
-                            }
-                            partitions.put(partitionColumns[ii], (Comparable<?>) value);
-                        }
-                        final IcebergTableLocationKey key =
-                                locationKey(df.format(), fileUri, partitions);
-                        // Verify before caching.
-                        return key.verifyFileReader() ? key : null;
-                    });
-                    if (locationKey != null) {
-                        locationKeyObserver.accept(locationKey);
-                    }
-                }
+
+        final PartitionData partitionData = (PartitionData) df.partition();
+        for (int ii = 0; ii < partitionColumns.length; ++ii) {
```

Comment:
We are driving the order from the table definition. There may be a mismatch here.